pytorch_lightning/overrides/data_parallel.py

```diff
@@ -194,6 +195,15 @@ def forward(self, *inputs, **kwargs):  # pragma: no-cover
             else:
                 output = self.module.validation_step(*inputs, **kwargs)
 
+        if not self._reducer_prepared_for_backwards:
+            self.reducer_prepare_for_backwards(output)
+
+        if output is None:
+            warn_missing_output(f'{fx_called} returned None. Did you forget to re')
+        return output
+
+    def reducer_prepare_for_backwards(self, output):
```

Comment:
If you look at the implementation of pytorch DDP https://pytorch.org/docs/stable/_modules/torch/nn/parallel/distributed.html#DistributedDataParallel
you can see they have a check 
`if torch.is_grad_enabled() and self.require_backward_grad_sync:`
so it's probably better to follow this implemenation and not to diverge even more. 

It would be nice to keep the implementations in sync #4630

Reply:
What I'm trying to say is can we just make use of the `self.require_backward_grad_sync` argument instead of introducing a new one?

Reply:
Thanks for the testing @awaelchli!

This unfortunately can't happen if we want to support manual optimization + ddp within the forward pass, things will have to change (since the behaviour is very different to DDP)... 

We could potentially add the backwards hook at the top of the forward, thoughts? I know of trying to reduce divergence, definitely think we can make it cleaner but we'll have to diverge further (wasn't too keen opening the PR for this, think we can do a few iterations on it).

Reply:
Yes I see `self.require_backward_grad_sync` is not what we need here I take that back.

Reply:
> We could potentially add the backwards hook at the top of the forward, thoughts?

That is a great idea. We should investigate that!

Reply:
Should we keep it as it is, and refactor this in a next PR, as currently manual_optimization is broken ?